### ISSUE TITLE
Content decoding failed.

### ISSUE THREAD

**Issue by MrSorcus** [mattermost/desktop#890]
I confirm (by marking "x" in the [ ] below: [x]):

- [ ] This is not a troubleshooting question. [Troubleshooting questions go here: http://www.mattermost.org/troubleshoot/](http://www.mattermost.org/troubleshoot/).
- [x] This doesn't reproduce on web browsers (such as in Chrome). If it does, [issue reports go to the Mattermost Server repository](https://github.com/mattermost/platform/issues).
- [x] I have read [contributing guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md).

---

https://mattermost.atlassian.net/browse/MM-14179

#### Summary
Desktop Mattermost doesn't work if all responses compressed with brotli.

#### Steps to reproduce
Mattermost-Desktop Version: 4.1.2

Enable brotli in the nginx config.
```
brotli on;
```

Try connect to server from mattermost-desktop.

#### Expected behavior
Mattermost-Desktop must works correctly with brotli compressed responses.

#### Observed behavior (that appears unintentional) 
```
errorCode: -330
errorDescription: "ERR_CONTENT_DECODING_FAILED"
type: "did-fail-load"
```

**Comment by amyblais** [comment]
@yuya-oc Wondering if this would be a feature request or a bug? 

I know we support webpack but not sure about brotli.

**Comment by yuya-oc** [comment]
@amyblais webpack is not related for this. It's about a way to decode contents of web servers.

The issue was not reproduced for me. The desktop app could show the content, but its encoding was gzip. On the other hand, Chrome can use brotli on the same server. In my investigation, the desktop app sends `Accept-Encoding: gzip, deflate` even when the server uses https.

- OS: macOS 10.14
- Client: Desktop app 4.1.2, Chrome 70.0.3538.77
- Server: nginx 1.14.0 (https://hub.docker.com/r/fholzer/nginx-brotli/) with my self-signed certificates.

I couldn't find significant issue other than https://github.com/electron/electron/issues/13647 and it suggests to customize `Accept-Encoding` header in our end. So I think the issue would be a feature request rather than a bug.

@MrSorcus Please let me confirm. Are any other compression methods completely disabled on your server?

**Comment by MrSorcus** [comment]
@yuya-oc yeah, only brotli enabled.
But this is can be reproduced only when nginx return (pre)compressed static files.
I mean when next scheme is used `Nginx (cache) <-> Nginx (compress static) <-> Mattermost`.
Where all requests from client to static going to Nginx (cache), then going to Nginx (compression), then going to Mattermost.
All other requests going to mattermost directly.
Yeah maybe that's looks ugly, but when the is no directly access to static resources, that's a good solution, where nginx can gave precompressed and cached static resources to client.

**Comment by MrSorcus** [comment]
Mobile app has similar behavior.

**Comment by yuya-oc** [comment]
Thanks for detailed description. Sadly I'm not familiar at complex nginx configuration though, I have some question.

- Where is `brotli on` configured on: `Nginx (cache)`, `Nginx (compress static)`, or both?
- What will happen when using the browser which doesn't support brotli? (for example, IE)

If you have a simple config file to reproduce the problem, it's very helpful for us.

**Comment by MrSorcus** [comment]
`Nginx Cache` as `NCache`. Used as Frontend web server.
`Nginx Compress` as `NCompress`. Used as compresing web server beetwen NCache and Mattermost.

Config for NCache.
```
worker_processes 1;
error_log        stderr warn;
pcre_jit         on;
user             nginx nginx;
pid              /run/nginx/nginx.pid;

events {
	worker_connections 1024;
	multi_accept       on;
}

http {
	charset UTF-8;
	default_type      application/octet-stream;
	keepalive_timeout 0;
	index             index.html;

	types_hash_max_size  3072;
	client_max_body_size 16m;

	server_tokens  off;
	log_subrequest off;
	log_format goaccess "[$time_local] $remote_addr "
		"\"$http_user_agent\" \"$http_referer\" "
		"\"$request\" $status $bytes_sent $request_time";
	access_log /home/access.log goaccess;

	proxy_buffering off;
	proxy_cache static;
	proxy_request_buffering off;
	proxy_cache_background_update on;
	proxy_cache_key "$request_method $http_host$uri";
	proxy_cache_valid 200 301 304 604800s;
	proxy_cache_path /tmp/cache levels=2:2 use_temp_path=off keys_zone=static:32m inactive=86400s;
	proxy_cache_use_stale updating;
	proxy_cache_lock on;

	sendfile                  on;
	tcp_nopush                on;
	tcp_nodelay               on;
	reset_timedout_connection on;

	ssl_protocols             TLSv1.2 TLSv1.3;
	ssl_ciphers               CHACHA20:AESGCM:!RSA;
	ssl_prefer_server_ciphers on;
	ssl_session_cache         shared:SSL:10m;
	ssl_session_timeout       10m;
	ssl_ecdh_curve            secp384r1:prime256v1;
	add_header                Strict-Transport-Security "max-age=15552000";

	include mime.types;

	# Force redirect all to https
	server {
		server_name .example.com;
		listen      139...:80;
		listen      [2a01:7e01:...]:80;
		return 301 https://$host$request_uri;
	}

	server {
		server_name chat.example.com;
		listen      139.162.170.136:443 ssl http2;
		listen      [2a01:7e01:e001:de:f96b:137f:131d:e26e]:443 ssl http2;

		ssl_certificate     /var/lib/acme.sh/example.com_ecc/fullchain.cer;
		ssl_certificate_key /var/lib/acme.sh/example.com_ecc/example.com.key;

		ssl_certificate     /var/lib/acme.sh/example.com/fullchain.cer;
		ssl_certificate_key /var/lib/acme.sh/example.com/example.com.key;

		location ~ \.(css|js|svg|woff2?)$ {
			proxy_cache static;
			proxy_buffering on;
			proxy_set_header Accept-Encoding "br";
			proxy_pass http://[2a01:7e01:...]:80;
		}

		location ~* /api/v[0-9]+/(users/)?websocket$ {
			proxy_http_version 1.1;
			proxy_set_header Upgrade "$http_upgrade";
			proxy_set_header Connection "upgrade";
			proxy_set_header Host "$http_host";
			proxy_pass http://[2a01:7e01:...]:8065;
		}

		location / {
			proxy_set_header Host "$http_host";
			proxy_pass http://[2a01:7e01:...]:8065;
		}
	}
}
```

Config for NCompress

```
worker_processes 1;
error_log        stderr warn;
pcre_jit         on;
user             nginx nginx;
pid              /run/nginx/nginx.pid;

events {
	worker_connections 1024;
	multi_accept       on;
}

http {
	charset UTF-8;
	default_type      application/octet-stream;
	keepalive_timeout 0;
	index             index.html;

	proxy_buffering         off;
	proxy_request_buffering off;

	types_hash_max_size  3072;
	client_max_body_size 16M;

	brotli            on;
	brotli_types      text/css application/javascript image/svg+xml text/javascript font/woff2 font/woff;
	brotli_comp_level 9;

	log_subrequest off;
	access_log     off;
	server_tokens  off;

	sendfile                  on;
	tcp_nopush                on;
	tcp_nodelay               on;
	aio                       on;
	reset_timedout_connection on;

	server {
		server_name chat.example.com;
		listen      [2a01:7e01:...]:80;

		location / {
			proxy_set_header Host $http_host;
			proxy_pass http://[2a01:7e01:...]:8065;
		}
	}

	include mime.types;
}
```

Now try to get some static resource.
`https://chat.example.com/static/main.58804551aaf9e4bce1d8.js` for example.
`curl https://chat.example.com/static/main.58804551aaf9e4bce1d8.js --output main.js.br`
Decompress it
`brotli -d main.js.br`

Now look at cache directory on NCache server
Resource is cached now as precompressed brotli file.
```
KEY: GET chat.example.com/static/main.58804551aaf9e4bce1d8.js
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 11 Nov 2018 14:48:32 GMT
Content-Type: application/javascript; charset=UTF-8
Connection: close
Cache-Control: max-age=31556926, public
Last-Modified: Wed, 24 Oct 2018 16:26:56 GMT
Vary: Accept-Encoding
Content-Encoding: br
```

**Comment by MrSorcus** [comment]
> Where is brotli on configured on: Nginx (cache), Nginx (compress static), or both?

Can be enabled on both. But i tested it with only on Nginx (compress static).
My latest config can check compression types and return cached precompressed static based on compress type. Look at the end for config (NCache).

> What will happen when using the browser which doesn't support brotli? (for example, IE)

I don't know. I use linux and here is no IE. But if you want, i can do check.
You can see usage statistics here - https://caniuse.com/#search=brotli

NCache with compress type checking (currently only uncompressed and brotli compressed static files)

```
worker_processes 1;
error_log        stderr warn;
pcre_jit         on;
user             nginx nginx;
pid              /run/nginx/nginx.pid;

events {
	worker_connections 1024;
	multi_accept       on;
}

http {
	charset UTF-8;
	default_type      application/octet-stream;
	keepalive_timeout 0;
	index             index.html;

	types_hash_max_size  3072;
	client_max_body_size 16m;

	server_tokens  off;
	log_subrequest off;
	log_format goaccess "[$time_local] $remote_addr "
		"\"$http_user_agent\" \"$http_referer\" "
		"\"$request\" $status $bytes_sent $request_time";
	access_log /home/access.log goaccess;

	map $http_accept_encoding $compress_type {
		"~*br" "br";
		default "";
	}

	proxy_buffering off;
	proxy_cache static;
	proxy_request_buffering off;
	proxy_cache_background_update on;
	proxy_cache_key "$request_method $compress_type $http_host$uri";
	proxy_cache_valid 200 301 304 604800s;
	proxy_cache_path /tmp/cache levels=2:2 use_temp_path=off keys_zone=static:32m inactive=86400s;
	proxy_cache_use_stale updating;
	proxy_cache_lock on;

	sendfile                  on;
	tcp_nopush                on;
	tcp_nodelay               on;
	reset_timedout_connection on;

	ssl_protocols             TLSv1.2 TLSv1.3;
	ssl_ciphers               CHACHA20:AESGCM:!RSA;
	ssl_prefer_server_ciphers on;
	ssl_session_cache         shared:SSL:10m;
	ssl_session_timeout       10m;
	ssl_ecdh_curve            secp384r1:prime256v1;
	add_header                Strict-Transport-Security "max-age=15552000";

	include mime.types;

	# Force redirect all to https
	server {
		server_name .example.com;
		listen      139...:80;
		listen      [2a01:7e01:...]:80;
		return 301 https://$host$request_uri;
	}

	server {
		server_name chat.example.com;
		listen      139...:443 ssl http2;
		listen      [2a01:7e01:...]:443 ssl http2;

		ssl_certificate     /var/lib/acme.sh/example.com_ecc/fullchain.cer;
		ssl_certificate_key /var/lib/acme.sh/example.com_ecc/example.com.key;

		ssl_certificate     /var/lib/acme.sh/example.com/fullchain.cer;
		ssl_certificate_key /var/lib/acme.sh/example.com/example.com.key;

		location ~ \.(css|js|svg|woff2?)$ {
			proxy_cache static;
			proxy_buffering on;
			proxy_set_header Accept-Encoding "$compress_type";
			proxy_pass http://[2a01:7e01:...]:80;
		}

		location ~* /api/v[0-9]+/(users/)?websocket$ {
			proxy_http_version 1.1;
			proxy_set_header Upgrade "$http_upgrade";
			proxy_set_header Connection "upgrade";
			proxy_set_header Host "$http_host";
			proxy_pass http://[2a01:7e01:...]:8065;
		}

		location / {
			proxy_set_header Host "$http_host";
			proxy_pass http://[2a01:7e01:...]:8065;
		}
	}
}
```
Now you can try to get resource with `Accept-Encoding: br` and without.

`curl --verbose https://chat.example.com/static/main.58804551aaf9e4bce1d8.js`
You receive uncompressed resource.

`curl --verbose https://chat.example.com/static/main.58804551aaf9e4bce1d8.js --header "Accept-Encoding: br"`
You receive brotli compressed resource.

In cache directory on NCache server you can see two difference files.
One is uncompressed
```
KEY: GET  chat.example.com/static/main.58804551aaf9e4bce1d8.js
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 11 Nov 2018 15:06:33 GMT
Content-Type: application/javascript; charset=UTF-8
Content-Length: 1163408
Connection: close
Accept-Ranges: bytes
Cache-Control: max-age=31556926, public
Last-Modified: Wed, 24 Oct 2018 16:26:56 GMT
Vary: Accept-Encoding
```
Second is brotli compressed
```
KEY: GET br chat.example.com/static/main.58804551aaf9e4bce1d8.js
HTTP/1.1 200 OK
Server: nginx
Date: Sun, 11 Nov 2018 15:06:44 GMT
Content-Type: application/javascript; charset=UTF-8
Connection: close
Cache-Control: max-age=31556926, public
Last-Modified: Wed, 24 Oct 2018 16:26:56 GMT
Vary: Accept-Encoding
Content-Encoding: br
```

**Comment by yuya-oc** [comment]
Thanks @MrSorcus! Assuming the later config, in my understanding, now you see static contents even when the client doesn't support brotli. Is that right?

By the way, I found Electron v4-beta sends `Accept-Encoding: gzip, deflate, br`. So we will be able to support brotli in the future version (Currently using Electron v2).

**Comment by amyblais** [comment]
Jira ticket for tracking: https://mattermost.atlassian.net/browse/MM-14179.

**Comment by yuya-oc** [comment]
Electron v4 was merged via #906. I believe this would be solved in v4.3.
